src/routes/votes.ts

```diff
@@ -0,0 +1,89 @@
+import * as Joi from '@hapi/joi';
+import { ContainerTypes, ValidatedRequest } from 'express-joi-validation';
+import { VOTES_TABLE, POSTS_TABLE, USER_HEADER } from '../constants';
+import * as pgsql from '../utils/pgsql';
+import { Post } from '../entities/post';
+import { BaseRequestSchema } from '../utils/schema';
+
+export const votesSchema = {
+	params: Joi.object({
+		post_id: Joi.string().required(),
+	}),
+};
+
+export interface VoteRequestSchema extends BaseRequestSchema {
+	[ContainerTypes.Params]: {
+		post_id: string;
+	};
+}
+
+export async function upvote(req: ValidatedRequest<VoteRequestSchema>) {
+	const vote = {
+		post_id: req.params.post_id,
+		user: req.headers[USER_HEADER],
+		vote: 1,
+	};
+
+	await saveVote(vote);
+}
+
+export async function downvote(req: ValidatedRequest<VoteRequestSchema>) {
+	const vote = {
+		post_id: req.params.post_id,
+		user: req.headers[USER_HEADER],
+		vote: -1,
+	};
+
+	await saveVote(vote);
+}
+
+type Vote = {
+	post_id: Post['id'];
+	user: string;
+	voted_at?: Date;
+	vote: number; // 1 = upvote, -1 = downvote
+};
+
+async function saveVote(vote: Vote) {
+	// save to votes table and update posts table within transaction
+	// ideally this should be done in a stored procedure
```

Comment:
This is a simple process composed of writing to a table, reading said table, performing simple calculations and updating a different table. Doing so in a stored procedure would be more efficient in so much as it would require only one connection and query to the DB .

Reply:
👍